source/common/network/address_impl.cc

```diff
@@ -8,32 +8,63 @@
 namespace Network {
 namespace Address {
 
-InstanceConstSharedPtr addressFromFd(int fd) {
-  sockaddr_storage ss;
-  socklen_t ss_len = sizeof ss;
-  const int rc = ::getsockname(fd, reinterpret_cast<sockaddr*>(&ss), &ss_len);
-  if (rc != 0) {
-    throw EnvoyException(fmt::format("getsockname failed for '{}': {}", fd, strerror(errno)));
-  }
+Address::InstanceConstSharedPtr addressFromSockAddr(const sockaddr_storage& ss, socklen_t ss_len) {
+  RELEASE_ASSERT(ss_len == 0 || ss_len >= sizeof(sa_family_t));
   switch (ss.ss_family) {
   case AF_INET: {
-    ASSERT(ss_len == sizeof(sockaddr_in));
+    RELEASE_ASSERT(ss_len == 0 || ss_len == sizeof(sockaddr_in));
     const struct sockaddr_in* sin = reinterpret_cast<const struct sockaddr_in*>(&ss);
     ASSERT(AF_INET == sin->sin_family);
     return InstanceConstSharedPtr(new Address::Ipv4Instance(sin));
```

Comment:
make_shared is almost always the right idea - single allocation, locality of reference, all that jazz.  I think there was some earlier question about whether the compiler can automatically substitute it but to my understanding this is not done.

The notable exception to it being the right idea is when weak_ptrs are in the picture.  As long as there are weak pointers left alive the control block must be kept around so divorcing the (small) control block from the (large) object probably makes sense for memory utilization reasons.